File: examples/game_development/tic_tac_toe.py
# @depyler: optimization_level = "aggressive"
# @depyler: bounds_checking = "explicit"
from typing import List, Optional, Tuple

class TicTacToe:
    """Tic-tac-toe game implementation"""
    
    def __init__(self) -> None:
        self.board: List[List[str]] = [
            [" ", " ", " "],
            [" ", " ", " "],
            [" ", " ", " "]
        ]
        self.current_player = "X"
    
    def make_move(self, row: int, col: int) -> bool:
        """Make a move at specified position"""
        if not self.is_valid_move(row, col):
            return False
        
        self.board[row][col] = self.current_player
        self.current_player = "O" if self.current_player == "X" else "X"
        return True
    
    def is_valid_move(self, row: int, col: int) -> bool:
        """Check if move is valid"""
        if row < 0 or row >= 3 or col < 0 or col >= 3:
            return False
        return self.board[row][col] == " "
    
    def check_winner(self) -> Optional[str]:
        """Check if there's a winner"""
        # Check rows
        for row in self.board:
            if row[0] == row[1] == row[2] != " ":
                return row[0]
        
        # Check columns
        for col in range(3):
            if self.board[0][col] == self.board[1][col] == self.board[2][col] != " ":
                return self.board[0][col]
        
        # Check diagonals
        if self.board[0][0] == self.board[1][1] == self.board[2][2] != " ":
            return self.board[0][0]
        
        if self.board[0][2] == self.board[1][1] == self.board[2][0] != " ":
            return self.board[0][2]
        
        return None
    
    def is_board_full(self) -> bool:
        """Check if board is full"""
        for row in self.board:
            for cell in row:
                if cell == " ":
                    return False
        return True
    
    def is_game_over(self) -> bool:
        """Check if game is over"""
        return self.check_winner() is not None or self.is_board_full()
    
    def get_empty_positions(self) -> List[Tuple[int, int]]:
        """Get all empty positions on the board"""
        positions: List[Tuple[int, int]] = []
        for row in range(3):
            for col in range(3):
                if self.board[row][col] == " ":
                    positions.append((row, col))
        return positions
    
    def board_to_string(self) -> str:
        """Convert board to string representation"""
        lines: List[str] = []
        for i, row in enumerate(self.board):
            line = f" {row[0]} | {row[1]} | {row[2]} "
            lines.append(line)
            if i < 2:
                lines.append("-----------")
        return "\n".join(lines)

class AIPlayer:
    """Simple AI player using minimax algorithm"""
    
    def __init__(self, symbol: str) -> None:
        self.symbol = symbol
        self.opponent = "O" if symbol == "X" else "X"
    
    def get_best_move(self, game: TicTacToe) -> Tuple[int, int]:
        """Get best move using minimax"""
        best_score = -1000
        best_move = (0, 0)
        
        for row, col in game.get_empty_positions():
            # Make temporary move
            game.board[row][col] = self.symbol
            score = self.minimax(game, 0, False)
            # Undo move
            game.board[row][col] = " "
            
            if score > best_score:
                best_score = score
                best_move = (row, col)
        
        return best_move
    
    def minimax(self, game: TicTacToe, depth: int, is_maximizing: bool) -> int:
        """Minimax algorithm implementation"""
        winner = game.check_winner()
        
        if winner == self.symbol:
            return 10 - depth
        elif winner == self.opponent:
            return depth - 10
        elif game.is_board_full():
            return 0
        
        if is_maximizing:
            best_score = -1000
            for row, col in game.get_empty_positions():
                game.board[row][col] = self.symbol
                score = self.minimax(game, depth + 1, False)
                game.board[row][col] = " "
                best_score = max(best_score, score)
            return best_score
        else:
            best_score = 1000
            for row, col in game.get_empty_positions():
                game.board[row][col] = self.opponent
                score = self.minimax(game, depth + 1, True)
                game.board[row][col] = " "
                best_score = min(best_score, score)
            return best_score

def play_computer_game() -> str:
    """Play a complete game against computer"""
    game = TicTacToe()
    ai = AIPlayer("O")
    moves_log: List[str] = []
    
    while not game.is_game_over():
        if game.current_player == "X":
            # Human player - make a simple move (first available)
            empty_positions = game.get_empty_positions()
            if empty_positions:
                row, col = empty_positions[0]
                game.make_move(row, col)
                moves_log.append(f"Human plays at ({row}, {col})")
        else:
            # AI player
            row, col = ai.get_best_move(game)
            game.make_move(row, col)
            moves_log.append(f"AI plays at ({row}, {col})")
    
    winner = game.check_winner()
    if winner:
        moves_log.append(f"Winner: {winner}")
    else:
        moves_log.append("Game ended in a tie")
    
    return "\n".join(moves_log)

def count_winning_positions(board_state: List[List[str]], player: str) -> int:
    """Count how many ways a player can win from current state"""
    winning_positions = [
        # Rows
        [(0, 0), (0, 1), (0, 2)],
        [(1, 0), (1, 1), (1, 2)],
        [(2, 0), (2, 1), (2, 2)],
        # Columns
        [(0, 0), (1, 0), (2, 0)],
        [(0, 1), (1, 1), (2, 1)],
        [(0, 2), (1, 2), (2, 2)],
        # Diagonals
        [(0, 0), (1, 1), (2, 2)],
        [(0, 2), (1, 1), (2, 0)]
    ]
    
    count = 0
    for positions in winning_positions:
        can_win = True
        for row, col in positions:
            cell = board_state[row][col]
            if cell != " " and cell != player:
                can_win = False
                break
        if can_win:
            count += 1
    
    return count

File: examples/game_development/tic_tac_toe.rs
#![allow(unused_imports)]
#![allow(unused_mut)]
#![allow(unused_variables)]
#![allow(unreachable_patterns)]
#![allow(unused_assignments)]
#![allow(dead_code)]
#[derive(Debug, Clone)]
pub struct IndexError {
    message: String,
}
impl std::fmt::Display for IndexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "index out of range: {}", self.message)
    }
}
impl std::error::Error for IndexError {}
impl IndexError {
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}
#[doc = r" Sum type for heterogeneous dictionary values(Python fidelity)"]
#[derive(Debug, Clone, PartialEq)]
pub enum DepylerValue {
    Int(i64),
    Float(f64),
    Str(String),
    Bool(bool),
    None,
    List(Vec<DepylerValue>),
    Dict(std::collections::HashMap<String, DepylerValue>),
}
impl std::fmt::Display for DepylerValue {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            DepylerValue::Int(i) => write!(f, "{}", i),
            DepylerValue::Float(fl) => write!(f, "{}", fl),
            DepylerValue::Str(s) => write!(f, "{}", s),
            DepylerValue::Bool(b) => write!(f, "{}", b),
            DepylerValue::None => write!(f, "None"),
            DepylerValue::List(l) => write!(f, "{:?}", l),
            DepylerValue::Dict(d) => write!(f, "{:?}", d),
        }
    }
}
impl DepylerValue {
    #[doc = r" Get length of string, list, or dict"]
    pub fn len(&self) -> usize {
        match self {
            DepylerValue::Str(s) => s.len(),
            DepylerValue::List(l) => l.len(),
            DepylerValue::Dict(d) => d.len(),
            _ => 0,
        }
    }
    #[doc = r" Check if empty"]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
    #[doc = r" Get chars iterator for string values"]
    pub fn chars(&self) -> std::str::Chars<'_> {
        match self {
            DepylerValue::Str(s) => s.chars(),
            _ => "".chars(),
        }
    }
    #[doc = r" Insert into dict(mutates self if Dict variant)"]
    pub fn insert(&mut self, key: String, value: DepylerValue) {
        if let DepylerValue::Dict(d) = self {
            d.insert(key, value);
        }
    }
    #[doc = r" Get value from dict by key"]
    pub fn get(&self, key: &str) -> Option<&DepylerValue> {
        if let DepylerValue::Dict(d) = self {
            d.get(key)
        } else {
            Option::None
        }
    }
    #[doc = r" Check if dict contains key"]
    pub fn contains_key(&self, key: &str) -> bool {
        if let DepylerValue::Dict(d) = self {
            d.contains_key(key)
        } else {
            false
        }
    }
    #[doc = r" Convert to String"]
    pub fn to_string(&self) -> String {
        match self {
            DepylerValue::Str(s) => s.clone(),
            DepylerValue::Int(i) => i.to_string(),
            DepylerValue::Float(fl) => fl.to_string(),
            DepylerValue::Bool(b) => b.to_string(),
            DepylerValue::None => "None".to_string(),
            DepylerValue::List(l) => format!("{:?}", l),
            DepylerValue::Dict(d) => format!("{:?}", d),
        }
    }
    #[doc = r" Convert to i64"]
    pub fn to_i64(&self) -> i64 {
        match self {
            DepylerValue::Int(i) => *i,
            DepylerValue::Float(fl) => *fl as i64,
            DepylerValue::Bool(b) => {
                if *b {
                    1
                } else {
                    0
                }
            }
            DepylerValue::Str(s) => s.parse().unwrap_or(0),
            _ => 0,
        }
    }
    #[doc = r" Convert to f64"]
    pub fn to_f64(&self) -> f64 {
        match self {
            DepylerValue::Float(fl) => *fl,
            DepylerValue::Int(i) => *i as f64,
            DepylerValue::Bool(b) => {
                if *b {
                    1.0
                } else {
                    0.0
                }
            }
            DepylerValue::Str(s) => s.parse().unwrap_or(0.0),
            _ => 0.0,
        }
    }
    #[doc = r" Convert to bool"]
    pub fn to_bool(&self) -> bool {
        match self {
            DepylerValue::Bool(b) => *b,
            DepylerValue::Int(i) => *i != 0,
            DepylerValue::Float(fl) => *fl != 0.0,
            DepylerValue::Str(s) => !s.is_empty(),
            DepylerValue::List(l) => !l.is_empty(),
            DepylerValue::Dict(d) => !d.is_empty(),
            DepylerValue::None => false,
        }
    }
}
impl std::ops::Index<usize> for DepylerValue {
    type Output = DepylerValue;
    fn index(&self, idx: usize) -> &Self::Output {
        match self {
            DepylerValue::List(l) => &l[idx],
            _ => panic!("Cannot index non-list DepylerValue"),
        }
    }
}
impl std::ops::Index<&str> for DepylerValue {
    type Output = DepylerValue;
    fn index(&self, key: &str) -> &Self::Output {
        match self {
            DepylerValue::Dict(d) => d.get(key).unwrap_or(&DepylerValue::None),
            _ => panic!("Cannot index non-dict DepylerValue with string key"),
        }
    }
}
#[derive(Debug, Clone)]
pub struct TicTacToe {
    pub board: Vec<Vec<String>>,
    pub current_player: String,
}
impl TicTacToe {
    pub fn new() -> Self {
        Self {
            board: Vec::new(),
            current_player: String::new(),
        }
    }
    pub fn make_move(&mut self, row: i32, col: i32) -> bool {
        if !self.is_valid_move(row, col) {
            return false;
        };
        self.board
            .clone()
            .get_mut(&row)
            .unwrap()
            .insert(col, self.current_player.clone());
        self.current_player = if self.current_player.clone() == "X".to_string() {
            "O".to_string()
        } else {
            "X".to_string()
        };
        return true;
    }
    pub fn is_valid_move(&self, row: i32, col: i32) -> bool {
        if row < 0 || row >= 3 || col < 0 || col >= 3 {
            return false;
        };
        return self.board.clone()[row as usize][col as usize] == " ".to_string();
    }
    pub fn check_winner(&self) -> Option<String> {
        for row in self.board.clone() {
            if row[0 as usize] == row[1 as usize]
                && row[1 as usize] == row[2 as usize]
                && row[2 as usize] != " ".to_string()
            {
                return Some(row[0 as usize]);
            };
        }
        for col in 0..3 {
            if self.board.clone()[0 as usize][col as usize]
                == self.board.clone()[1 as usize][col as usize]
                && self.board.clone()[1 as usize][col as usize]
                    == self.board.clone()[2 as usize][col as usize]
                && self.board.clone()[2 as usize][col as usize] != " ".to_string()
            {
                return Some(self.board.clone()[0 as usize][col as usize]);
            };
        }
        if self.board.clone()[0 as usize][0 as usize] == self.board.clone()[1 as usize][1 as usize]
            && self.board.clone()[1 as usize][1 as usize]
                == self.board.clone()[2 as usize][2 as usize]
            && self.board.clone()[2 as usize][2 as usize] != " ".to_string()
        {
            return Some(self.board.clone()[0 as usize][0 as usize]);
        };
        if self.board.clone()[0 as usize][2 as usize] == self.board.clone()[1 as usize][1 as usize]
            && self.board.clone()[1 as usize][1 as usize]
                == self.board.clone()[2 as usize][0 as usize]
            && self.board.clone()[2 as usize][0 as usize] != " ".to_string()
        {
            return Some(self.board.clone()[0 as usize][2 as usize]);
        };
        return None;
    }
    pub fn is_board_full(&self) -> bool {
        for row in self.board.clone() {
            for cell in row {
                if cell == " ".to_string() {
                    return false;
                };
            }
        }
        return true;
    }
    pub fn is_game_over(&self) -> bool {
        return self.check_winner().is_some() || self.is_board_full();
    }
    pub fn get_empty_positions(&self) -> Vec<(i32, i32)> {
        let mut positions = vec![];
        for row in 0..3 {
            for col in 0..3 {
                if self.board.clone()[row as usize][col as usize] == " ".to_string() {
                    positions.push((row, col));
                };
            }
        }
        return positions;
    }
    pub fn board_to_string(&self) -> String {
        let mut lines = vec![];
        for (i, row) in self
            .board
            .clone()
            .iter()
            .cloned()
            .enumerate()
            .map(|(i, x)| (i as i32, x))
        {
            let line = format!(
                " {} | {} | {} ",
                row[0 as usize], row[1 as usize], row[2 as usize]
            );
            lines.push(line);
            if i < 2 {
                lines.push("-----------".to_string());
            };
        }
        return lines.join("\n".to_string());
    }
}
#[derive(Debug, Clone)]
pub struct AIPlayer {
    pub symbol: String,
    pub opponent: DepylerValue,
}
impl AIPlayer {
    pub fn new(symbol: String) -> Self {
        Self {
            symbol,
            opponent: Default::default(),
        }
    }
    pub fn get_best_move(&self, game: &TicTacToe) -> (i32, i32) {
        let mut best_score = -1000;
        let mut best_move = (0, 0);
        for (row, col) in game.get_empty_positions() {
            game.board
                .get_mut(&row)
                .unwrap()
                .insert(col, self.symbol.clone());
            let score = self.minimax(game, 0, false);
            game.board
                .get_mut(&row)
                .unwrap()
                .insert(col, " ".to_string());
            if score > best_score {
                let best_score = score;
                let best_move = (row, col);
            };
        }
        return best_move;
    }
    pub fn minimax(&self, game: &TicTacToe, depth: i32, is_maximizing: bool) -> i32 {
        let winner = game.check_winner();
        if winner == self.symbol.clone() {
            return 10 - depth;
        } else {
            if winner == self.opponent.clone() {
                return depth - 10;
            } else {
                if game.is_board_full() {
                    return 0;
                };
            };
        };
        if is_maximizing {
            let mut best_score = -1000;
            for (row, col) in game.get_empty_positions() {
                game.board
                    .get_mut(&row)
                    .unwrap()
                    .insert(col, self.symbol.clone());
                let score = self.minimax(game, depth + 1, false);
                game.board
                    .get_mut(&row)
                    .unwrap()
                    .insert(col, " ".to_string());
                let best_score = (best_score).max(score);
            }
            return best_score;
        } else {
            let mut best_score = 1000;
            for (row, col) in game.get_empty_positions() {
                game.board
                    .get_mut(&row)
                    .unwrap()
                    .insert(col, self.opponent.clone());
                let score = self.minimax(game, depth + 1, true);
                game.board
                    .get_mut(&row)
                    .unwrap()
                    .insert(col, " ".to_string());
                let best_score = (best_score).min(score);
            }
            return best_score;
        };
    }
}
#[doc = "Play a complete game against computer"]
pub fn play_computer_game() -> Result<String, Box<dyn std::error::Error>> {
    let mut game = TicTacToe::new();
    let ai = AIPlayer::new("O".to_string());
    let mut moves_log: Vec<String> = vec![];
    while !game.is_game_over() {
        let mut col;
        let mut row;
        if game.current_player == "X" {
            let empty_positions = game.get_empty_positions();
            if empty_positions {
                (row, col) = empty_positions
                    .get(0usize)
                    .cloned()
                    .expect("IndexError: list index out of range");
                game.make_move(row, col);
                moves_log.push(format!("Human plays at({:?}, {:?})", row, col));
            }
        } else {
            (row, col) = ai.get_best_move(&game);
            game.make_move(row, col);
            moves_log.push(format!("AI plays at({:?}, {:?})", row, col));
        }
    }
    let winner = game.check_winner();
    if winner {
        moves_log.push(format!("Winner: {:?}", winner));
    } else {
        moves_log.push("Game ended in a tie".to_string().to_string());
    }
    Ok(moves_log.join("\n"))
}
#[doc = "Count how many ways a player can win from current state"]
pub fn count_winning_positions<'b, 'a>(
    board_state: &'a Vec<Vec<String>>,
    player: &'b str,
) -> Result<i32, Box<dyn std::error::Error>> {
    let mut count: i32 = Default::default();
    let winning_positions = vec![
        vec![(0, 0), (0, 1), (0, 2)],
        vec![(1, 0), (1, 1), (1, 2)],
        vec![(2, 0), (2, 1), (2, 2)],
        vec![(0, 0), (1, 0), (2, 0)],
        vec![(0, 1), (1, 1), (2, 1)],
        vec![(0, 2), (1, 2), (2, 2)],
        vec![(0, 0), (1, 1), (2, 2)],
        vec![(0, 2), (1, 1), (2, 0)],
    ];
    count = 0;
    for positions in winning_positions.iter().cloned() {
        let mut can_win = true;
        for (row, col) in positions.iter().cloned() {
            let cell = board_state
                .get(row as usize)
                .cloned()
                .expect("IndexError: list index out of range")
                .get(col as usize)
                .cloned()
                .expect("IndexError: list index out of range");
            if (cell != " ") && (cell != player) {
                can_win = false;
                break;
            }
        }
        if can_win {
            count = count + 1;
        }
    }
    Ok(count)
}
